accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,156 @@
+
+        public void Add(string key, int value) { }
+        public void Add(string key, double value) { }
+        public void Add(string key, bool value) { }
+        public void Add(string key, string value) { }
+        public void Add(string key, int [] value) { }
+        public void Add(string key, double [] value) { }
+        public void Add(string key, bool [] value) { }
+        public void Add(string key, string [] value) { }
+        public void Add(string key, ActivityTagsCollection value) { }
+
+        // interfaces implementation
+        public int Count { get; }
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator() { ... }
+        IEnumerator IEnumerable.GetEnumerator() { ... }
+    }
+```
+
+## Activity.Tags
+
+As mentioned earlier, Activity has Tags property which returns the list of the activity tags. Tags are a list of `KeyValuePair<string, string>`. Tags are proven to be not enough for OT scenarios as we need to support more value types (numeric, bool, numeric array, and nested tags). Activity today has the method `public Activity AddTag(string key, string? value)`. The proposal here is to add more `AddTag` overload accepting more types and add a new getter property return the list of all typed tags. 
+
+```C#
+    public partial class Activity : IDisposable
+    {
+        public void AddTag(string key, int value) { }
+        public void AddTag(string key, double value) { }
+        public void AddTag(string key, bool value) { }
+        public void AddTag(string key, int [] value) { }
+        public void AddTag(string key, double [] value) { }
+        public void AddTag(string key, bool [] value) { } // how interesting this one?
+        public void AddTag(string key, string [] value) { }
+        public void AddTag(string key, ActivityTagsCollection value) { }
+
+        // We cannot change Tags property so we are introducing a new property with a new name
+        public ActivityTagsCollection TypedTags { get; }
+    }
+```
+
+## Proposal Changing already exposed APIs in previous preview releases
+
+As we are introducing the ActivityTagsCollection, we can take advantage of that now and modify some of the exposed APIs to use this collection instead. 
+
+### ActivityEvent
+
+```C#
+    // Already exposed APIs
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityTagsCollection? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityTagsCollection? attributes)
```

Comment:
Just curious (probably not related to this PR), would `attributes` get deep-copied or a shallow copy? (e.g. if the tags collection has reference to another tags collection)

Reply:
This is needed to be decided which behavior is the correct one. I'll try to look at OT specs to know if it mentions the expected behavior.

Reply:
I am wondering, do we need to do a copy at all? I understand the specs saying Event is immutable but does that need to copy (either shallow or deep) the attributes? what python or other implementations do in such cases? I know nested attributes are not approved/implemented yet but I am wondering in general about the current state.

Reply:
This is decided to not doing any copies and instead the created event object will just hold on the passed input collection.